### PR TITLE
Better search ui

### PR DESCRIPTION
**Pull request by core23**
This will update the ui of the search result page:
* Result page is better integrated in the admin theme
* Added icon to result list
* Mobile page breaks on different screen sizes
 

Before:
bildschirmfoto 2015-09-09 um 11 56 12

After:
bildschirmfoto 2015-09-09 um 11 55 21

### CODE REVIEW

**Comment by Soullivaneuh** [comment]
Looks nice. :+1: 

Please fix the tests.

File: Resources/views/Block/block_search_result.html.twig [hunk 1]
```diff
@@ -12,43 +12,52 @@ file that was distributed with this source code.
...
+            <div class="panel-heading">
+                <h4 class="panel-title">
+                    {% set icon = settings.icon|default('') %}
+                    {{ icon|raw }}
+
+                    {% if admin.hasRoute('list') %}
+                        <a href="{{ admin.generateUrl('list') }}">{{ admin.label|trans({}, admin.translationdomain) }}</a>
                     {% else %}
-                        <li><i>{{ admin.toString(result) }}</i></li>
+                        {{ admin.label|trans({}, admin.translationdomain) }}
```
Comment (Soullivaneuh) [inline Resources/views/Block/block_search_result.html.twig]: Why removing the `toString` method?

**Comment by Soullivaneuh** [comment]
Please use AdminLTE theme boxes: https://almsaeedstudio.com/themes/AdminLTE/pages/UI/general.html

**Comment by Soullivaneuh** [comment]
Maybe full width table would give a better rendering: https://almsaeedstudio.com/themes/AdminLTE/pages/tables/simple.html

**Comment by Soullivaneuh** [comment]
Could you please take a look at overflow management?

search_box_overflow

File: Resources/views/Block/block_search_result.html.twig [hunk 2]
```diff
@@ -12,43 +12,52 @@ file that was distributed with this source code.
...
-                {% endif %}
-            </h4>
-        </div>
-
-        <ul>
-            {% if pager %}
-                {% for result in pager.getResults() %}
-                    {% if admin.hasRoute('edit') %}
-                        <li><a href="{{ admin.generateObjectUrl('edit', result) }}">{{ admin.toString(result) }}</a></li>
+    <div class="col-md-4 col-sm-6">
```
Comment (Soullivaneuh) [inline Resources/views/Block/block_search_result.html.twig]: `col-lg-4 col-md-6` would be clearer.

**Comment by rande** [comment]
+1

**Comment by TomasVotruba** [comment]
@core23 This is really great job, thank you! The page is now very nice and readable.

**Comment by OskarStark** [comment]
@core23 nice!!

File: Resources/views/Block/block_search_result.html.twig [hunk 3]
```diff
@@ -12,43 +12,52 @@ file that was distributed with this source code.
...
+                        {% if pager.getNbResults() > pager.getResults()|length %}
+                            <span class="list-group-item"><i class="fa fa-ellipsis-h"></i></span>
+                        {% endif %}
                     {% else %}
-                        <li><i>{{ 'no_results_found'|trans({}, 'SonataAdminBundle') }}</i></li>
+                        {% if admin.hasRoute('create') %}
+                            <span>{{ 'no_results_found'|trans({}, 'SonataAdminBundle') }}</span>
+                        {% else %}
+                            <span>{{ 'no_results_found'|trans({}, 'SonataAdminBundle') }}</span>
+                        {% endif %}
```
Comment (mvhirsch) [inline Resources/views/Block/block_search_result.html.twig]: This looks redundant to me: if hasRoute() then 'no_results_found' else 'no_results_found'.
Did you forget something or can we reduce this to one ```<span>```?

Comment (Soullivaneuh) [inline Resources/views/Block/block_search_result.html.twig]: +1, create link is missing.

File: Resources/views/Core/search.html.twig [hunk 4]
```diff
@@ -31,32 +31,29 @@ file that was distributed with this source code.
...
+        {% for group in groups %}
+            {% set display = (group.roles is empty or is_granted('ROLE_SUPER_ADMIN') ) %}
+            {% for role in group.roles if not display %}
+                {% set display = is_granted(role) %}
             {% endfor %}
+
+            {% if display %}
+                {% for admin in group.items %}
+                    {% set count = count + 1 %}
+                    {% if admin.hasroute('create') and admin.isGranted('CREATE') or admin.hasroute('list') and admin.isGranted('LIST') %}
```
Comment (mvhirsch) [inline Resources/views/Core/search.html.twig]: Please use ```hasRoute``` (uppercase R) instead of ```hasroute```.

**Comment by mvhirsch** [comment]
+1, nice work @core23

File: Resources/views/Block/block_search_result.html.twig [hunk 5]
```diff
@@ -12,43 +12,52 @@ file that was distributed with this source code.
...
+                            {% endif %}
+                        {% else %}
+                            {% if admin.hasRoute('create') %}
+                                <span>{{ 'no_results_found'|trans({}, 'SonataAdminBundle') }}</span>
+                            {% else %}
+                                <span>{{ 'no_results_found'|trans({}, 'SonataAdminBundle') }}</span>
+                            {% endif %}
+                        {% endfor %}
+                        {% if pager.getNbResults() > pager.getResults()|length %}
+                            <span class="list-group-item"><i class="fa fa-ellipsis-h"></i></span>
```
Comment (Soullivaneuh) [inline Resources/views/Block/block_search_result.html.twig]: Not sure if this ellipsis is very needed. This add a new line without link, not really useful IMO.

**Comment by core23** [comment]
Thank you for the great feedback :)

Here is a new draft
bildschirmfoto 2015-09-09 um 15 01 05

File: Resources/views/Block/block_search_result.html.twig [hunk 6]
```diff
@@ -12,43 +12,46 @@ file that was distributed with this source code.
...
+                        {% for result in pager.getResults() %}
+                            {% if admin.hasRoute('edit') %}
+                                <li><a href="{{ admin.generateObjectUrl('edit', result) }}">{{ admin.toString(result) }}</a></li>
+                            {% else %}
+                                <li><a>{{ admin.toString(result) }}</a></li>
+                            {% endif %}
+                        {% endfor %}
+                        {% if pager.getNbResults() > pager.getResults()|length %}
+                            {% if admin.hasRoute('list') %}
+                                <li><a href="{{ admin.generateUrl('list') }}"><i class="fa fa-ellipsis-h"></i></a></li>
```
Comment (Soullivaneuh) [inline Resources/views/Block/block_search_result.html.twig]: Same here. Don't know if just a list link is useful.

@rande What do you think?

File: Resources/views/Block/block_search_result.html.twig [hunk 7]
```diff
@@ -12,43 +12,46 @@ file that was distributed with this source code.
...
-                    {% else %}
-                        <li><i>{{ 'no_results_found'|trans({}, 'SonataAdminBundle') }}</i></li>
-                    {% endif %}
-                {% endfor %}
+            </div>
+            {% if pager and pager.getResults()|length %}
+                <div class="box-body no-padding">
+                    <ul class="nav nav-stacked">
+                        {% for result in pager.getResults() %}
+                            {% if admin.hasRoute('edit') %}
```
Comment (Soullivaneuh) [inline Resources/views/Block/block_search_result.html.twig]: Edit grant should be checked.

File: Resources/views/Block/block_search_result.html.twig [hunk 8]
```diff
@@ -12,43 +12,46 @@ file that was distributed with this source code.
...
             {% else %}
-                {% if admin.hasRoute('create') %}
-                    <li><i>{{ 'no_results_found'|trans({}, 'SonataAdminBundle') }} ~ <a href="{{ admin.generateUrl('create') }}">{{ 'add_new_entry'|trans({}, 'SonataAdminBundle') }}</a></i></li>
-                {% else %}
-                    <li><i>{{ 'no_results_found'|trans({}, 'SonataAdminBundle') }}</i></li>
-                {% endif %}
+                <div class="box-body">
+                    <p>
+                        <em>{{ 'no_results_found'|trans({}, 'SonataAdminBundle') }}</span></em>
+                        {% if admin.hasRoute('create') %}
```
Comment (Soullivaneuh) [inline Resources/views/Block/block_search_result.html.twig]: Create grant should be checked.

File: Resources/views/Block/block_search_result.html.twig [hunk 9]
```diff
@@ -12,43 +12,46 @@ file that was distributed with this source code.
...
+                    <ul class="nav nav-stacked">
+                        {% for result in pager.getResults() %}
+                            {% if admin.hasRoute('edit') %}
+                                <li><a href="{{ admin.generateObjectUrl('edit', result) }}">{{ admin.toString(result) }}</a></li>
+                            {% else %}
+                                <li><a>{{ admin.toString(result) }}</a></li>
+                            {% endif %}
+                        {% endfor %}
+                        {% if pager.getNbResults() > pager.getResults()|length %}
+                            {% if admin.hasRoute('list') %}
```
Comment (Soullivaneuh) [inline Resources/views/Block/block_search_result.html.twig]: List grant should be checked.

File: Resources/views/Block/block_search_result.html.twig (hunk 6 above)
Comment (rande) [inline Resources/views/Block/block_search_result.html.twig]: Yes, maybe a link with filters already set make more sense

Comment (Soullivaneuh) [inline Resources/views/Block/block_search_result.html.twig]: Yes, but we don't have a global filter on admin list. :-)

**Comment by core23** [comment]
Maybe the add new entry link could be placed in the header? What do you think?

bildschirmfoto 2015-09-09 um 15 32 27

**Comment by OskarStark** [comment]
:+1:  for new entry Link in header

what do you think that we add some configuration options for the search results?

**Comment by Soullivaneuh** [comment]
Agree for the header link @core23.

> what do you think that we add some configuration options for the search results?

@OskarStark Which configuration? Could be done on another PR after this one IMO.

**Comment by OskarStark** [comment]
:+1:  for separate PR

@config values: for example ```create_link_in_header``` (naming is just an example)

**Comment by Soullivaneuh** [comment]
@OskarStark Better place for template override IMHO.

**Comment by OskarStark** [comment]
:+1:  this should work to, but the it should be a small block to avoid duplication of many code

File: tmppP7EMg [hunk 10]
```diff
@@ -0,0 +1,33 @@
+<?php
```
Comment (Soullivaneuh) [inline tmppP7EMg]: Hmm what is this? :-)

Comment (core23) [inline tmppP7EMg]: no idea ^^

File: Resources/views/Block/block_search_result.html.twig [hunk 11]
```diff
@@ -12,43 +12,44 @@ file that was distributed with this source code.
...
+        <div class="box">
+            <div class="box-header with-border">
+                <h3 class="box-title">
+                    {% set icon = settings.icon|default('') %}
+                    {{ icon|raw }}
                     {{ admin.label|trans({}, admin.translationdomain) }}
-                {% endif %}
-
+                </h3>
                 {% if pager and pager.getNbResults() > 0 %}
```
Comment (Soullivaneuh) [inline Resources/views/Block/block_search_result.html.twig]: According to [AdminLTE documentation](https://almsaeedstudio.com/themes/AdminLTE/documentation/index.html#component-box), it should be:

```twig
<div class="box-tools pull-right">
    {% if pager and pager.getNbResults() > 0 %}
        <span class="badge">{{ pager.getNbResults() }}</span>
    {% elseif admin.hasRoute('create') and admin.isGranted('CREATE') %}
        <a href="{{ admin.generateUrl('create') }}" class="btn btn-box-tool">
            <i class="fa fa-plus"></i>
        </a>
    {% endif %}
</div>
```

And I think create button could be here even if search has some results.

File: Resources/views/Block/block_search_result.html.twig [hunk 12]
```diff
@@ -12,43 +12,44 @@ file that was distributed with this source code.
...
+                    <ul class="nav nav-stacked">
+                        {% for result in pager.getResults() %}
+                            {% if admin.hasRoute('edit') and admin.isGranted('EDIT', result) %}
+                                <li><a href="{{ admin.generateObjectUrl('edit', result) }}">{{ admin.toString(result) }}</a></li>
+                            {% else %}
+                                <li><a>{{ admin.toString(result) }}</a></li>
+                            {% endif %}
+                        {% endfor %}
+                        {% if pager.getNbResults() > pager.getResults()|length %}
+                            {% if admin.hasRoute('list') and admin.isGranted('LIST') %}
```
Comment (Soullivaneuh) [inline Resources/views/Block/block_search_result.html.twig]: Please remember https://github.com/sonata-project/SonataAdminBundle/pull/3235#discussion-diff-39039285

**Comment by Soullivaneuh** [comment]
Missing overflow management.

File: Resources/views/Block/block_search_result.html.twig [hunk 13]
```diff
@@ -12,43 +12,44 @@ file that was distributed with this source code.
...
-        <div>
-            <h4>
-                {% if admin.hasRoute('list') %}
-                    <a href="{{ admin.generateUrl('list') }}">{{ admin.label|trans({}, admin.translationdomain) }}</a>
-                {% else %}
+    <div class="col-lg-4 col-md-6">
+        <div class="box">
+            <div class="box-header with-border">
+                <h3 class="box-title">
+                    {% set icon = settings.icon|default('') %}
```
Comment (Soullivaneuh) [inline Resources/views/Block/block_search_result.html.twig]: Icon must be outside `.box-title`.

File: Resources/views/Core/search.html.twig [hunk 14]
```diff
@@ -25,38 +25,35 @@ file that was distributed with this source code.
     </div>
     -->
 
-    <h1>{{ 'title_search_results'|trans({'%query%': query}, 'SonataAdminBundle') }}</h1>
+    <h2 class="page-header">{{ 'title_search_results'|trans({'%query%': query}, 'SonataAdminBundle') }}</h2>
```
Comment (Soullivaneuh) [inline Resources/views/Core/search.html.twig]: Please follow bootstrap documentation: http://getbootstrap.com/components/#page-header

File: Resources/views/Block/block_search_result.html.twig [hunk 15]
```diff
@@ -12,43 +12,44 @@ file that was distributed with this source code.
...
+                </div>
             {% else %}
-                {% if admin.hasRoute('create') %}
-                    <li><i>{{ 'no_results_found'|trans({}, 'SonataAdminBundle') }} ~ <a href="{{ admin.generateUrl('create') }}">{{ 'add_new_entry'|trans({}, 'SonataAdminBundle') }}</a></i></li>
-                {% else %}
-                    <li><i>{{ 'no_results_found'|trans({}, 'SonataAdminBundle') }}</i></li>
-                {% endif %}
+                <div class="box-body">
+                    <p>
+                        <em>{{ 'no_results_found'|trans({}, 'SonataAdminBundle') }}</span></em>
```
Comment (Soullivaneuh) [inline Resources/views/Block/block_search_result.html.twig]: From where `</span>` is coming from?

**Comment by core23** [comment]
[**@Soullivaneuh**](https://github.com/Soullivaneuh) commented on [10. Sep. 2015 13:28 MESZ](https://github.com/sonata-project/SonataAdminBundle/pull/3235#issuecomment-139209072 "2015-09-10T11:28:45Z - Replied by Github Comment Enhancer"):
> Missing overflow management. [#3235 (comment)](https://github.com/sonata-project/SonataAdminBundle/pull/3235#issuecomment-138889216 "Better search ui")

This is a bug in the AdminLTE theme which should be fixed there.

**Comment by Soullivaneuh** [comment]
> This is a bug in the AdminLTE theme which should be fixed there.

Where?

**Comment by Soullivaneuh** [comment]
I don't think it's a bug, this is the default behavior in bootstrap.

But in this case, we need to manage overflow.

A single `word-break: break-word;` on css file should do the trick.

File: Resources/views/Block/block_search_result.html.twig (hunk 13 above)
Comment (core23) [inline Resources/views/Block/block_search_result.html.twig]: Where? Can't find any example in the other block templates.

Comment (Soullivaneuh) [inline Resources/views/Block/block_search_result.html.twig]: Looks at code of first box here: https://almsaeedstudio.com/themes/AdminLTE/pages/UI/general.html

File: Resources/public/css/layout.css [hunk 16]
```diff
@@ -355,6 +355,10 @@ body.fixed .content-header .navbar.stuck {
     border-radius: 0;
 }
 
+.box .nav-stacked > li {
```
Comment (Soullivaneuh) [inline Resources/public/css/layout.css]: Not sure if it should be on `layout.css` or `styles.css`. @rande can you explain the difference?

And a special css class would be better IMO.

File: Resources/views/Block/block_search_result.html.twig (hunk 13 above)
Comment (core23) [inline Resources/views/Block/block_search_result.html.twig]: Thanks :)

File: Resources/public/css/layout.css (hunk 16 above)
Comment (core23) [inline Resources/public/css/layout.css]: As mentioned before, this is a general bug in the nav-stacked element. So why use a special css class?

Comment (Soullivaneuh) [inline Resources/public/css/layout.css]: As I said, I don't think it's a bug, this is the default behavior in bootstrap.

Overflow management may be decided by us, not by bootstrap.

I don't know if apply it on generic navs would not make some side effects.

@rande what do you think?

File: Resources/views/Block/block_search_result.html.twig [hunk 17]
```diff
@@ -12,43 +12,48 @@ file that was distributed with this source code.
...
-                        <li><a href="{{ admin.generateObjectUrl('edit', result) }}">{{ admin.toString(result) }}</a></li>
-                    {% else %}
-                        <li><i>{{ admin.toString(result) }}</i></li>
+    <div class="col-lg-4 col-md-6">
+        <div class="box">
+            <div class="box-header with-border">
+                {{ icon|raw }}
+                {{ admin.label|trans({}, admin.translationdomain) }}
+                <h3 class="box-title">
+                    {% set icon = settings.icon|default('') %}
```
Comment (Soullivaneuh) [inline Resources/views/Block/block_search_result.html.twig]: I think you made a mistake here. icon variable should be before icon rendering and admin label on title, isn't it?

Comment (core23) [inline Resources/views/Block/block_search_result.html.twig]: Sorry, was a stupid copy & paste fault

**Comment by OskarStark** [comment]
any news on this PR?

**Comment by Soullivaneuh** [comment]
@OskarStark waiting https://github.com/sonata-project/SonataAdminBundle/pull/3235#discussion_r39249202 answer.

File: Resources/public/css/layout.css (hunk 16 above)
Comment (rande) [inline Resources/public/css/layout.css]: The original idea behind ``layout.css`` and ``styles.css`` was (5 years ago ;) to allow overwrite colors without breaking the structure defined in layout.

@Soullivaneuh maybe we should fix for the search view only by adding a specific class

Comment (Soullivaneuh) [inline Resources/public/css/layout.css]: Ok so it should be on `layout.css`.

> maybe we should fix for the search view only by adding a specific class

As I was thinking. @core23 could you please do that?

Comment (core23) [inline Resources/public/css/layout.css]: Added the class `sonata-search-result-list` for the search result. Hope your now fine with it ;)

**Comment by Soullivaneuh** [comment]
@core23 missing https://github.com/sonata-project/SonataAdminBundle/pull/3235/files#r39149241

At least, you can removed the ellipsis and put back list link on box title or widget icon.

**Comment by core23** [comment]
When putting a link to the headline, the text would become blue to indicate a link. This would break the general styling of boxes in the existing admin backend. We could remove the color, but the ui would break, because links have differnt colors :/
bildschirmfoto 2015-09-26 um 13 32 26

Another idea is putting the ellipsis to the head
bildschirmfoto 2015-09-26 um 13 30 43

**Comment by Soullivaneuh** [comment]
> Another idea is putting the ellipsis to the head

Nice idea, but I would change the icon. Propositions:

* http://fontawesome.io/icon/arrow-right/
* http://fontawesome.io/icon/arrow-circle-right/

**Comment by core23** [comment]
@Soullivaneuh What about the list icon? It is used in any other context to return to the admin lists.
https://fortawesome.github.io/Font-Awesome/icon/list/

**Comment by Soullivaneuh** [comment]
@core23 Indeed, didn't think about this one.

Let's go for list icon so. :+1:

**Comment by OskarStark** [comment]
@core23 can you provide a screenshot with the new list icon?

**Comment by Soullivaneuh** [comment]
@core23 list icon looks very small. Could you please add `fa-lg` class on it?

**Comment by core23** [comment]
Normal icons
bildschirmfoto 2015-09-30 um 18 09 35
Large icons
bildschirmfoto 2015-09-30 um 18 09 09

**Comment by Soullivaneuh** [comment]
Well, let's keep normal icon. It's better to keep theme sizes.

**Comment by OskarStark** [comment]
@core23 can you please change the icon again?

**Comment by core23** [comment]
@OskarStark Which icon do you mean?

**Comment by OskarStark** [comment]
> Well, let's keep normal icon. It's better to keep theme sizes.

**Comment by core23** [comment]
The icon should be normal, as I did not create any new commit ;)

**Comment by OskarStark** [comment]
Ah ok, didn't notice :smile: 

@Soullivaneuh can we merge this PR then?

**Comment by Soullivaneuh** [comment]
Finally! :-)

Thank you for your contribution @core23. :+1:

**Comment by rande** [comment]
GG ;)

**Comment by OskarStark** [comment]
Hooray :smile:

**Comment by core23** [comment]
:smiley: Thanks!

**Pull request merged**